sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/interface.scala

```diff
@@ -183,9 +183,11 @@ case class CatalogTable(
 
   import CatalogTable._
 
-  /** schema of this table's partition columns */
-  def partitionSchema: StructType = StructType(schema.filter {
-    c => partitionColumnNames.contains(c.name)
+  /** schema of this table's partition columns
+   * keep the schema order with partitionColumnNames
+   */
+  def partitionSchema: StructType = StructType(partitionColumnNames.flatMap {
+    p => schema.filter(_.name == p)
```

Comment:
I think this is a bug, can you send an individual PR for this bug? I'd like to backport this bug fix to 2.1